library/ssl_tls13_generic.c

```diff
@@ -296,51 +308,32 @@ static int ssl_tls13_parse_certificate_verify( mbedtls_ssl_context *ssl,
     p += 2;
     MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, signature_len );
 
-    /* Hash verify buffer with indicated hash function */
-    switch( md_alg )
+    status = psa_hash_compute( hash_alg,
+                               verify_buffer,
+                               verify_buffer_len,
+                               verify_hash,
+                               sizeof( verify_hash ),
+                               &verify_hash_len );
+    if( status != PSA_SUCCESS )
     {
-#if defined(MBEDTLS_SHA256_C)
-        case MBEDTLS_MD_SHA256:
-            verify_hash_len = 32;
-            ret = mbedtls_sha256( verify_buffer, verify_buffer_len, verify_hash, 0 );
-            break;
-#endif /* MBEDTLS_SHA256_C */
-
-#if defined(MBEDTLS_SHA384_C)
-        case MBEDTLS_MD_SHA384:
-            verify_hash_len = 48;
-            ret = mbedtls_sha512( verify_buffer, verify_buffer_len, verify_hash, 1 );
-            break;
-#endif /* MBEDTLS_SHA384_C */
-
-#if defined(MBEDTLS_SHA512_C)
-        case MBEDTLS_MD_SHA512:
-            verify_hash_len = 64;
-            ret = mbedtls_sha512( verify_buffer, verify_buffer_len, verify_hash, 0 );
-            break;
-#endif /* MBEDTLS_SHA512_C */
-
-        default:
-            ret = MBEDTLS_ERR_SSL_HANDSHAKE_FAILURE;
-            break;
+        MBEDTLS_SSL_DEBUG_RET( 1, "hash computation PSA error", status );
+        goto error;
     }
 
-    if( ret != 0 )
+    MBEDTLS_SSL_DEBUG_BUF( 3, "verify hash", verify_hash, verify_hash_len );
+
+    const mbedtls_md_info_t* md_info;
+    md_info = mbedtls_md_info_from_psa( hash_alg );
+    if( md_info == NULL )
     {
-        MBEDTLS_SSL_DEBUG_RET( 1, "hash computation error", ret );
-        goto error;
+        return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
     }
+    mbedtls_md_type_t md_alg = mbedtls_md_get_type( md_info );
```

Comment:
Some compilers complain that you're redefining `md_alg` that was already define before and I think they're right. Also, it was already set on line 277 so doesn't need to be set here again - I think this is a leftover from the previous version that didn't use `mbedtls_ssl_tls13_get_pk_type_and_md_alg_from_sig_alg()`, but now that we do, I think this paragraph can be removed. (And then the block below for `RSA_PSS` can be left as it was before this commit.)

Reply:
(Note: `md_info` needs to remain in the `RSA_PSS` block otherwise in builds without RSA-PSS we get a warning about it being unused.)

Also: changing the `RSA_PSS` block will be the topic of #5796 - hence out of scope for this PR.